logger.go

```diff
@@ -297,11 +297,17 @@ func (log *Logger) check(lvl zapcore.Level, msg string) *zapcore.CheckedEntry {
 	// Thread the error output through to the CheckedEntry.
 	ce.ErrorOutput = log.errorOutput
 	if log.addCaller {
-		ce.Entry.Caller = zapcore.NewEntryCaller(runtime.Caller(log.callerSkip + callerSkipOffset))
-		if !ce.Entry.Caller.Defined {
+		pc := make([]uintptr, 1)
+		numFrames := runtime.Callers(log.callerSkip+callerSkipOffset+1, pc)
+		frame, _ := runtime.CallersFrames(pc[:numFrames]).Next()
+		defined := frame.PC != 0
+		if !defined {
 			fmt.Fprintf(log.errorOutput, "%v Logger.check error: failed to get caller\n", time.Now().UTC())
 			log.errorOutput.Sync()
 		}
+
+		ce.Entry.Caller = zapcore.NewEntryCaller(frame.PC, frame.File, frame.Line, defined)
+		ce.Entry.Caller.Function = frame.Function
```

Comment:
since we're setting fields anyway, we could just do `zapcore.EntryCaller{PC: frame.PC, File: frame.File, ...}` instead of using the constructor but then modifying (prefer to create once and use that instead of mutating the `Entry.Caller`)

Reply:
thanks @prashantv updated